fabtests/component/dmabuf-rdma/fi-rdmabw-rocr.c

```diff
@@ -0,0 +1,1282 @@
+
+	sprintf(port_name, "%d", port);
+	version = FI_VERSION(1, 12);
+	if (ep_type == FI_EP_MSG) {
+		if (client)
+			EXIT_ON_ERROR(fi_getinfo(version, server_name,
+						 port_name, 0, hints, &fi));
+		else
+			EXIT_ON_ERROR(fi_getinfo(version, server_name,
+						 port_name, FI_SOURCE, hints,
+						 &fi));
+	} else {
+		EXIT_ON_ERROR(fi_getinfo(version, NULL, NULL, 0, hints, &fi));
+	}
+
+	fi_freeinfo(hints);
+
+	if (ep_type == FI_EP_RDM || client)
+		printf("Using OFI device: %s (%s)\n",
+			fi->fabric_attr->prov_name, fi->domain_attr->name);
+
+	EXIT_ON_ERROR(fi_fabric(fi->fabric_attr, &fabric, NULL));
+	if (ep_type == FI_EP_MSG) {
+		EXIT_ON_ERROR(fi_eq_open(fabric, &eq_attr, &eq, NULL));
+		if (!client) {
+			fi_pep = fi;
+			EXIT_ON_ERROR(fi_passive_ep(fabric, fi_pep, &pep, NULL));
+			EXIT_ON_ERROR(fi_pep_bind(pep, (fid_t)eq, 0));
+			EXIT_ON_ERROR(fi_listen(pep));
+			EXIT_ON_ERROR(wait_conn_req(eq, &fi));
+			printf("Using OFI device: %s (%s)\n",
+				fi_pep->fabric_attr->prov_name,
+				fi->domain_attr->name);
+		}
+	}
+	EXIT_ON_ERROR(fi_domain(fabric, fi, &domain, NULL));
+	EXIT_ON_ERROR(fi_endpoint(domain, fi, &ep, NULL));
+	EXIT_ON_ERROR(fi_cq_open(domain, &cq_attr, &cq, NULL));
+	if (ep_type == FI_EP_RDM) {
+		EXIT_ON_ERROR(fi_av_open(domain, &av_attr, &av, NULL));
+		EXIT_ON_ERROR(fi_ep_bind(ep, (fid_t)av, 0));
+	} else {
+		EXIT_ON_ERROR(fi_ep_bind(ep, (fid_t)eq, 0));
+	}
+	EXIT_ON_ERROR(fi_ep_bind(ep, (fid_t)cq,
+		    (FI_TRANSMIT | FI_RECV | FI_SELECTIVE_COMPLETION)));
+	EXIT_ON_ERROR(fi_enable(ep));
+
+	if (ep_type == FI_EP_MSG) {
+		if (client)
+			EXIT_ON_ERROR(fi_connect(ep, fi->dest_addr, NULL, 0));
+		else
+			EXIT_ON_ERROR(fi_accept(ep, NULL, 0));
+		EXIT_ON_ERROR(wait_connected(ep, eq));
+	}
+
+	if (test_type == SEND &&
+	    !(fi->domain_attr->mr_mode & (FI_MR_HMEM | FI_MR_LOCAL))) {
+		printf("Local MR registration skipped.\n");
+		goto done;
+	}
+
+	if (fi->domain_attr->mr_mode & FI_MR_RAW)
+		use_raw_key = 1;
+
+	for (i = 0; i < num_gpus; i++) {
+		iov.iov_base = bufs[i].rocr_buf.buf;
+		iov.iov_len = max_size * batch;
+		mr_attr.mr_iov = &iov;
+		mr_attr.iov_count = 1;
+		mr_attr.access = FI_REMOTE_READ | FI_REMOTE_WRITE |
+				 FI_READ | FI_WRITE | FI_SEND | FI_RECV;
+		mr_attr.requested_key = i + 1;
+		mr_attr.iface = bufs[i].rocr_buf.location == MALLOC ?
+					FI_HMEM_SYSTEM : FI_HMEM_ROCR;
+		EXIT_ON_ERROR(fi_mr_regattr(domain, &mr_attr, 0, &mr));
+
+		if (fi->domain_attr->mr_mode & FI_MR_ENDPOINT) {
+			EXIT_ON_ERROR(fi_mr_bind(mr, (fid_t)ep, 0));
+			EXIT_ON_ERROR(fi_mr_enable(mr));
+		}
+
+		bufs[i].mrs[nic] = mr;
+	}
+
+	if (buf_location == DEVICE && use_proxy) {
+		iov.iov_base = proxy_buf.rocr_buf.buf;
+		iov.iov_len = max_size * batch;
+		mr_attr.mr_iov = &iov;
+		mr_attr.iov_count = 1;
+		mr_attr.access = FI_REMOTE_READ | FI_REMOTE_WRITE |
+				 FI_READ | FI_WRITE | FI_SEND | FI_RECV;
+		mr_attr.requested_key = i + 1;
+        //Check these
+		//mr_attr.device.ze = rocr_get_dev_num(i);
```

Comment:
does `rocr_get_dev_num()` work?

Reply:
no, for above reason I didnt implement one. Even with device selection (once we figure that out), I don't think we ever have need for this funciton?